sys/boot/geli/geliboot.c

```diff
@@ -27,53 +27,75 @@
  * $FreeBSD$
  */
 
-#include <crypto/intake.h>
-
+#include "geliboot_internal.h"
 #include "geliboot.h"
 
 SLIST_HEAD(geli_list, geli_entry) geli_head = SLIST_HEAD_INITIALIZER(geli_head);
 struct geli_list *geli_headp;
 
-typedef u_char geli_mkey[G_ELI_DATAIVKEYLEN];
+typedef u_char geli_ukey[G_ELI_USERKEYLEN];
 
-static geli_mkey saved_keys[GELI_MAX_KEYS];
+static geli_ukey saved_keys[GELI_MAX_KEYS];
 static unsigned int nsaved_keys = 0;
 
+/*
+ * Copy keys from local storage to the keybuf struct.
+ * Destroy the local storage when finished.
+ */
 void
-geli_fill_keybuf(struct keybuf *keybuf)
+geli_fill_keybuf(struct keybuf *fkeybuf)
 {
-        unsigned int i;
+	unsigned int i;
 
-        for (i = 0; i < nsaved_keys; i++) {
-                keybuf->kb_ents[i].ke_type = KEYBUF_TYPE_GELI;
-                memcpy(keybuf->kb_ents[i].ke_data, saved_keys[i],
-                    G_ELI_DATAIVKEYLEN);
-        }
+	for (i = 0; i < nsaved_keys; i++) {
+		fkeybuf->kb_ents[i].ke_type = KEYBUF_TYPE_GELI;
+		memcpy(fkeybuf->kb_ents[i].ke_data, saved_keys[i],
+		    G_ELI_USERKEYLEN);
+	}
+	fkeybuf->kb_nents = nsaved_keys;
+	bzero(saved_keys, sizeof(saved_keys));
+}
 
-        keybuf->kb_nents = nsaved_keys;
-        bzero(saved_keys, sizeof(saved_keys));
+/*
+ * Copy keys from a keybuf struct into local storage.
+ * Zero out the keybuf.
+ */
+void
+geli_save_keybuf(struct keybuf *skeybuf)
```

Comment:
This function is used to fill the local variable from a keybuf passed from boot2